IdentityCore/src/workplacejoin/ios/MSIDWorkPlaceJoinUtil.m

```diff
@@ -105,19 +109,84 @@ + (MSIDRegistrationInformation *)getRegistrationInformation:(id<MSIDRequestConte
     return info;
 }
 
-+ (SecIdentityRef)copyWPJIdentity:(__unused id<MSIDRequestContext>)context
++ (SecIdentityRef)copyWPJIdentityWithTenantId:(NSString *)tenantId
+                                      context:(id<MSIDRequestContext>)context
+                            certificateIssuer:(NSString **)issuer
+                               privateKeyDict:(NSDictionary **)keyDict
+{
+    NSString *teamId = [[MSIDKeychainUtil sharedInstance] teamId];
+    
+    if (!teamId)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Encountered an error when reading teamID from keychain.");
+        return nil;
+    }
+    
+    NSString *legacySharedAccessGroup = [NSString stringWithFormat:@"%@.com.microsoft.workplacejoin", teamId];
+    
+    SecIdentityRef legacyIdentity = [self copyWPJIdentity:context tag:nil sharedAccessGroup:legacySharedAccessGroup certificateIssuer:issuer privateKeyDict:keyDict];
+    
+    if (legacyIdentity)
+    {
+        if ([NSString msidIsStringNilOrBlank:tenantId])
+        {
+            // ESTS didn't request a specific tenant, just return default one
+            return legacyIdentity;
+        }
+        
+        // Read tenantId for legacy identity
+        NSString *registrationTenantId = [MSIDWorkPlaceJoinUtil getWPJStringDataForIdentifier:kMSIDTenantKeyIdentifier context:context error:nil];
+        
+        // There's no tenantId on the registration, or it mismatches what server requested, keep looking for a better match. Otherwise, return the identity already.
+        if (![NSString msidIsStringNilOrBlank:registrationTenantId]
+            && [registrationTenantId isEqualToString:tenantId])
+        {
+            return legacyIdentity;
+        }
+    }
+    
+    NSString *defaultSharedAccessGroup = [NSString stringWithFormat:@"%@.com.microsoft.workplacejoin.v2", teamId];
+    NSString *tag = [NSString stringWithFormat:@"%@#%@", kWPJPrivateKeyIdentifier, tenantId];
+    
+    SecIdentityRef secondaryIdentity = [self copyWPJIdentity:context
+                                                         tag:tag
+                                           sharedAccessGroup:defaultSharedAccessGroup
+                                           certificateIssuer:issuer
+                                              privateKeyDict:keyDict];
+    
+    // If secondary Identity was found, return it
+    if (secondaryIdentity)
+    {
+        return secondaryIdentity;
+    }
+    
+    // Otherwise, return legacy Identity - this can happen if we couldn't match based on the tenantId, but Identity was there. It could be usable. We'll let ESTS to evaluate it and check.
+    // This means that for registrations that have no tenantId stored, we'd always do this extra query until registration gets updated to have the tenantId stored on it.
+    return legacyIdentity;
```

Comment:
This line can also be hit if both legacyIdentity and secondaryIdentity are nil. Is it okay to return nil?

Reply:
Yes, it's okay